src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -66,73 +91,98 @@ private static string GetMetadataNameWithoutBackticks(MetadataReader reader, Str
-        private static async ValueTask<SymbolTreeInfo> GetInfoForMetadataReferenceAsync(
+        public static async ValueTask<SymbolTreeInfo> GetInfoForMetadataReferenceAsync(
             Solution solution,
             PortableExecutableReference reference,
-            Checksum checksum,
+            Checksum? checksum,
             CancellationToken cancellationToken)
         {
-            var metadataId = GetMetadataIdNoThrow(reference);
-            if (metadataId == null)
-                return CreateEmpty(checksum);
+            checksum ??= GetMetadataChecksum(solution.Services, reference, cancellationToken);
 
-            if (s_metadataIdToInfo.TryGetValue(metadataId, out var infoTask))
+            if (s_peReferenceToInfo.TryGetValue(reference, out var infoTask))
             {
                 var info = await infoTask.GetValueAsync(cancellationToken).ConfigureAwait(false);
-                if (info.Checksum == checksum)
-                    return info;
+                Contract.ThrowIfFalse(info.Checksum != checksum, "How could the info stored for a particular PEReference now have a different checksum?");
+                return info;
             }
 
-            var metadata = GetMetadataNoThrow(reference);
-            if (metadata == null)
-                return CreateEmpty(checksum);
-
             return await GetInfoForMetadataReferenceSlowAsync(
-                solution.Services, SolutionKey.ToSolutionKey(solution), reference, metadata, cancellationToken).ConfigureAwait(false);
+                solution.Services, SolutionKey.ToSolutionKey(solution), reference, checksum, cancellationToken).ConfigureAwait(false);
+
+            static async Task<SymbolTreeInfo> GetInfoForMetadataReferenceSlowAsync(
+                SolutionServices services,
+                SolutionKey solutionKey,
+                PortableExecutableReference reference,
+                Checksum checksum,
+                CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Important: this captured async lazy may live a long time *without* computing the final results. As
+                // such, it is important that it not capture any large state.  For example, it should not hold onto a
+                // Solution instance.
+                //
+                // this is keyed per reference, so that have unique SymbolTreeInfo's per reference with their own
+                // correct checksum.  Ensuring we only compute this once per *Metadata* instance though is handled below in 
+                // CreateMetadataSymbolTreeInfoAsync
+                var asyncLazy = s_peReferenceToInfo.GetValue(
+                    reference,
+                    id => new AsyncLazy<SymbolTreeInfo>(
+                        c => CreateMetadataSymbolTreeInfoAsync(services, solutionKey, reference, checksum, c),
+                        cacheResult: true));
```

Comment:
this cached from `PERef -> SymbolTreeInfo`.

Reply:
i still may have a followup that actually splits the type into just the raw data, and then the raw-data+checksum.  but i considered that too large to do here.

Reply:
I like the idea, that `WithChecksum` bit does feel a little hacky :)